### ISSUE TITLE
How to dispatch on molar mass?

### ISSUE THREAD

**Issue by ejmeitz** [PainterQubits/Unitful.jl#662]
Hello,

I'd like to differentiate my functions based on if the user provided a `Mass` dimension or a molar mass dimension. Unitful already has the `Unitful.Mass` to make it easy to dispatch on mass but I've been majorly struggling to dispatch on molar mass. My attempt is below:

```julia
molar_mass_unit = Unitful.Dimensions{(Unitful.Dimension{:Mass}(1//1), Unitful.Dimension{:Amount}(-1//1))}
function f(a::Unitful.Quantity{T, molar_mass_unit, U}) where {T,U}
       println("molar mass function")
end
```
I also tried using 
```julia
 molar_mass_unit = Unitful.Dimensions{(Unitful.𝐌,Unitful.𝐍^-1)}
```

**Comment by ejmeitz** [comment]
Ok like 50th thing I tried worked but I still don't get why the above doesn't

Using: `mm_dim = dimension(u"g/mol")` as the dimension in the quantity above works.

**Comment by sostock** [comment]
> My attempt is below:
> ```julia
> molar_mass_unit = Unitful.Dimensions{(Unitful.Dimension{:Mass}(1//1), Unitful.Dimension{:Amount}(-1//1))}
> function f(a::Unitful.Quantity{T, molar_mass_unit, U}) where {T,U}
>        println("molar mass function")
> end
> ```

This doesn’t work for two reasons:
1. The second type parameter must be an _instance_ of a `Dimensions` type, not a type itself (i.e., `molar_mass_unit()` instead of `molar_mass_unit`)
2. Unitful uses canonical ordering of the dimensions in the `Dimensions` type parameter. That means that a value of `1u"kg/mol"` always has dimension

> I also tried using
> ```julia
> molar_mass_unit = Unitful.Dimensions{(Unitful.𝐌,Unitful.𝐍^-1)}
> ```

This doesn’t work because `Unitful.𝐌` and `Unitful.𝐍` are already instances of the type `Unitful.Dimensions`, not `Unitful.Dimension`. The following works:
```julia
molar_mass_unit = Unitful.𝐌 * Unitful.𝐍^-1
```

**Comment by sostock** [comment]
> My attempt is below:
> ```julia
> molar_mass_unit = Unitful.Dimensions{(Unitful.Dimension{:Mass}(1//1), Unitful.Dimension{:Amount}(-1//1))}
> function f(a::Unitful.Quantity{T, molar_mass_unit, U}) where {T,U}
>        println("molar mass function")
> end
> ```

This doesn’t work for two reasons:
1. The second type parameter must be an _instance_ of a `Dimensions` type, not a type itself (i.e., `molar_mass_unit()` instead of `molar_mass_unit`)
2. Unitful uses a canonical ordering of the dimensions in the `Dimensions` type parameter. A value of `1u"kg/mol"` (and any other quantity or unit of molar mass) has dimension
   ```julia
   julia> typeof(dimension(1u"kg/mol"))
   Unitful.Dimensions{(Unitful.Dimension{:Amount}(-1//1), Unitful.Dimension{:Mass}(1//1))}
   ```
   As you can see, the order of the dimensions is switched compared to your attempt. That’s why it doesn’t work.

> I also tried using
> ```julia
> molar_mass_unit = Unitful.Dimensions{(Unitful.𝐌,Unitful.𝐍^-1)}
> ```

This doesn’t work because `Unitful.𝐌` and `Unitful.𝐍` are already instances of the type `Unitful.Dimensions`, not `Unitful.Dimension`. The following works:
```julia
molar_mass_unit = Unitful.𝐌 * Unitful.𝐍^-1
```

**Comment by sostock** [comment]
We should have better documentation for this. I think we should

1. Feature `@derived_dimension` more prominently in the manual.
2. Improve the part of the `Quantity` docstring which explains that `D` is an instance while `U` is a type:
   > The type parameters `D :: Unitful.Dimensions` and `U <: Unitful.Units`.

**Comment by ejmeitz** [comment]
Yeah definitely add another example to the dispatch part. Its a super useful and cool feature.
